File: data/gencode/SkillCfg.go
package gencode

import (
	"encoding/json"
	"gs/define"
	"gs/lib/mylog"
	"io/ioutil"
	"os"
	"sync"
)

type SkillCfg struct {
	SkillSlc []*SkillInfo `json:"Skill"`
	SkillMap map[int32]*SkillInfo
}

type SkillInfo struct {
	ID            int32  // 技能id
	Name          string // 技能名称
	Level         int32  // 技能等级
	BeforeTime    int32  // 前摇时间ms
	AfterTime     int32  // 后摇时间ms
	IsWeaponSkill bool   // 是否为武器平A技能
	CoolDown      int32  // 冷却时间ms
	TargetType    int32  // 目标选择类型
}

var skillCfg *SkillCfg
var skillOnce sync.Once

func GetSkillCfg() *SkillCfg {
	skillOnce.Do(func() {
		skillCfg = new(SkillCfg)
		skillCfg.init()
	})
	return skillCfg
}

func (this *SkillCfg) init() {
	this.SkillMap = make(map[int32]*SkillInfo)
	rootPath := os.Getenv(define.EnvName)
	filePtr, err := os.Open(rootPath + "/data/json/Skill.json")
	if err != nil {
		mylog.Error("load SkillCfg failed", err)
		return
	}
	defer filePtr.Close()
	data, err := ioutil.ReadAll(filePtr)
	if err != nil {
		mylog.Error("load SkillCfg failed", err)
		return
	}
	err = json.Unmarshal(data, this)
	if err != nil {
		mylog.Error("load SkillCfg failed", err)
		return
	}
	for _, v := range this.SkillSlc {
		this.SkillMap[v.ID] = v
	}
}

func (this *SkillCfg) GetSkillById(id int32) (*SkillInfo, bool) {
	if ret, ok := this.SkillMap[id]; ok {
		return ret, ok
	}
	return nil, false
}

File: data/gencode/SkillCfg.ts
import Res from "../../common/util/Res"

export class SkillInfo {
    public ID: number// 技能id
    public Name: string// 技能名称
    public Level: number// 技能等级
    public BeforeTime: number// 前摇时间ms
    public AfterTime: number// 后摇时间ms
    public IsWeaponSkill: boolean// 是否为武器技能
    public CoolDown: number// 冷却时间ms
    public TargetType: number// 目标选择类型
    public Attack: number// 伤害量
    public Heal: number// 治疗量
}

export class SkillConfig {
    private static instance: SkillConfig
    public static Get(): SkillConfig {
        if (SkillConfig.instance == null) {
            SkillConfig.instance = new SkillConfig()
            SkillConfig.instance.init()
        }
        return SkillConfig.instance
    }

    public SkillSlc: Array<SkillInfo>
    public SkillMap: Map<number, SkillInfo>

    private init(): void {
        this.SkillSlc = new Array<SkillInfo>()
        this.SkillMap = new Map<number, SkillInfo>()

        let jsonData = Res.get<cc.JsonAsset>("json/Skill", cc.JsonAsset)
        this.SkillSlc = jsonData.json['Skill']
        this.SkillSlc.forEach(skill => {
            this.SkillMap.set(skill.ID, skill)
        })

    }
}
